### PR TITLE
Implement gallery POTD in Home view

### PR DESCRIPTION
**Pull request by bgerstle**
Phab: [T119669](https://phabricator.wikimedia.org/T119669), [T119903](https://phabricator.wikimedia.org/T119903)

- [x] refactor galleries
  - [x] base gallery (RTL handling + data source property)
  - [x] article header gallery (basic data source for article images)
  - [x] modal article gallery (modal data source for article images + info)
  - [x] modal POTD gallery (modal data source for POTD images + info)
- [x] tap POTD to view in gallery
- [x] peek POTD to view in gallery
- [x] fetch "medium" POTD image w/ desc for home view
- [x] fetch "large" POTD image w/ all info for modal gallery
- [x] **extra credit** add older POTDs in gallery

regression tests: (see Phab tasks for dev testing notes)

---

### Image gallery VC & data source architecture
See header comments in:

- `WMFModalImageGalleryViewController`: formerly known as `WMFImageGalleryViewController`, which is now more dependent on its data sources (i.e. no longer fetches image info).
- `WMFImageGalleryDataSource`: and its "big brother,"`WMFModalImageGalleryDataSource` are the interface through which domain-specific data can be displayed in the image gallery:
  - `WMFModalArticleImageGalleryDataSource`: as the name implies, provides data to the modal gallery for images in an article
  - `WMFModalPOTDImageGalleryDataSource`: I think you've got it by now: this one provides data for the gallery when viewing pictures of the day

Each gallery data source is created as part of one of `WMFModalImageGalleryViewController`'s new initialization methods.  Check the header for details.

### Standardized image widths
Took a little time to refactor all our disparate image-width stuff into a centralized collection of categories. See `UIScreen+WMFImageWidth` for details.  
#### Takeaways
- they're collected in a `UIScreen` category since all the view-specific resolutions are dependent on the screen's scale
- we're maxing out scaling at 2x to prevent ourselves from fetching insanely high resolution images for the home feed.  For example: 3x scaling would result in 960-px thumbnails (if available) for an article preview cell in the home view :scream:
- we were previously _always_ showing 1280px thumbnails in the gallery, it now scales from 640 to 1280 depending on screen scale. this means older devices will get lower-res gallery pics and newer devices will remain the same. 
- ideally, we'd download higher resolution images on 2-3x devices when on WiFi, but that's out of scope for now. I'll file a ticket to look into this later.

### Split `WMFHomeSectionController` protocols
As part of adding POTD to the Home view, we needed a way to handle selection of home section items which weren't related to articles.  And lo, two new protocols were conceived:

- `WMFArticleHomeSectionController`: all the article-specific stuff from `WMFHomeSectionController`
- `WMFGenericHomeSectionController`: (open to better names) non-article stuff for home sections. only provides a "detail view controller" when one of its items are selected. might need to add specialized footer interaction methods here if we ever get around to doing a "more POTDs" list view.

## Add'l boyscouting
- Removed unused gallery & gallery cell properties (most of these were added when I tried to use `WMFImageGalleryViewController` for _all_ galleries, but that got way too complicated, and I forgot to remove these properties when I backed out the changes)
- DRY'd up image view placeholder logic in the cells (category already existed it, just integrated in more places and provided explicit, documented overrides where necessary)
- Fixed image gallery animations. Previously the detail overlay and higher-res images weren't animating into view, nor was the chrome cross dissolving on taps

### CODE REVIEW

**Comment by montehurd** [comment]
Looks like this also fixes https://phabricator.wikimedia.org/T118927 :)

**Comment by bgerstle** [comment]
ah the 10m timeout strikes again! let's see if the latest commit on #298 passes. if so maybe we can merge it and see if this will pass after merging master

**Comment by bgerstle** [comment]
:-1: will address @coreyfloyd's comment by trying to consolidate `WMFModalImageGalleryViewController` subclasses using custom initializers if possible (should be possible).

**Comment by bgerstle** [comment]
:checkered_flag: @coreyfloyd should be ready for re-review

**Comment by bgerstle** [comment]
after that last refactor, need run a couple more regression tests (specifically lead image tap while article downloads on slow connection). everything else should be good (checked header/modal gallery image sync already)

**Comment by bgerstle** [comment]
@coreyfloyd @montehurd this takes test coverage down a wee bit.  I'd like to do a follow-on for adding visual tests for cell placeholder styling, but I could add unit tests for the date stuff I added, and _maybe_ some dataSource logic. Some of this stuff is a bit harder to test at a low level, and probably best left for user acceptance tests.

**Comment by montehurd** [comment]
IT'S BEAUTIFUL! :)

The only thing I'd add is the date in the gallery view - that is, a title saying:

"Picture of the Day"
        12/3/15

Gives it context and rewards the author a bit by not only having their name at the bottom by the CC, but also by saying, "hey, this pic is the freaking pic of the day!" :)

File: Wikipedia/Code/WMFArticleListTableViewCell.xib [hunk 1]
```diff
@@ -10,7 +10,7 @@
         <tableViewCell contentMode="scaleToFill" selectionStyle="default" indentationWidth="10" rowHeight="60" id="KGk-i7-Jjw" customClass="WMFArticleListTableViewCell">
             <rect key="frame" x="0.0" y="0.0" width="458" height="60"/>
             <autoresizingMask key="autoresizingMask" flexibleMaxX="YES" flexibleMaxY="YES"/>
-            <tableViewCellContentView key="contentView" opaque="NO" clipsSubviews="YES" multipleTouchEnabled="YES" contentMode="center" tableViewCell="KGk-i7-Jjw" id="H2p-sc-9uM">
+            <tableViewCellContentView key="contentView" opaque="NO" clipsSubviews="YES" multipleTouchEnabled="YES" contentMode="center" ambiguous="YES" tableViewCell="KGk-i7-Jjw" id="H2p-sc-9uM">
```
Comment (montehurd) [inline Wikipedia/Code/WMFArticleListTableViewCell.xib]: ambiguous!

**Comment by bgerstle** [comment]
Yeah, the header isn't even up to spec. Maybe we can work the date in there
to match featured article, @nirzar?

**Comment by montehurd** [comment]
@bgerstle Just confirmed with both Kaity and Josh that adding "Picture of the Day: 12/3/15" as the first line of the description would be great for now :)

**Comment by bgerstle** [comment]
K, will add that in

**Comment by montehurd** [comment]
Just in the gallery view. The feed view is cool as-is.

**Comment by bgerstle** [comment]
@montehurd ah, that's different... Follow on since it wasn't in original
spec? Current architecture renders both galleries identically. Adding
custom UI would take a bit more work. Not much, but would probably be
better spent getting current scoped beta stuff in (like compass fix)

**Comment by montehurd** [comment]
No custom UI needed I think... just tack it on the the beginning of the descrip text:
screen shot 2015-12-03 at 4 42 31 pm

**Comment by bgerstle** [comment]
I'll look into it, but alright if this is merged without this?

**Comment by montehurd** [comment]
For sure!

**Comment by montehurd** [comment]
:+1:

**Comment by montehurd** [comment]
Oh did you see the ambiguous constraint comment I left?

**Comment by bgerstle** [comment]
@montehurd in the feed:

image


in the gallery:

image

**Comment by bgerstle** [comment]
hm, there's something weird going on w/ the truncation now. will investigate...

**Comment by bgerstle** [comment]
fixed truncation:

portrait: 

image


landscape:

image

**Comment by coreyfloyd** [comment]
:+1:

**Pull request merged**